lantiq: use nvmem for bthomehub v5a

neheb opened the pull request:
Userspace handling of both calibration and mac addresses is deprecated.

Also fixed calibration size for ath9k. AR9287 uses 3d8 for its size.

ping @xdarklight 

@DragonBluep it's claimed to use AR9287 but the PCI ID claims AR9227. I assume they are equivalent.

neheb committed:
lantiq: use nvmem for bthomehub v5a

Userspace handling of both calibration and mac addresses is deprecated.

Also fixed calibration size for ath9k. AR9287 uses 3d8 for its size.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17289
Signed-off-by: Robert Marko <robimarko@gmail.com>

xdarklight commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_bt_homehub-v5a.dts:
```diff
@@ -262,12 +294,20 @@
-	wifi@168c,002d {
+	wifi@1 {
 		compatible = "pci168c,002d";
-		reg = <0x7000 0 0 0 0>;
```
is this `reg` change on purpose? if so it should be documented in the commit description

neheb replied:
Might be a mistake. I don’t remember if I observed dtc warnings.

neheb replied:
Removed

xdarklight commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_bt_homehub-v5a.dts:
```diff
@@ -262,12 +294,20 @@
 		ieee80211-freq-limit = <2402000 2482000>;
 	};
+
+	wifi@2 {
+		compatible = "qcom,ath10k";
+		reg = <0x2000 0 0 0 0>;
+		nvmem-cells = <&cal_caldata_5000>, <&macaddr_caldata_110c 3>;
+		nvmem-cell-names = "calibration", "mac-address";
+	};
```
this is not correct, the ath10k card is on the PCIe bus (`&pcie0`)

please note that the PCIe driver unfortunately is an out-of-tree one and I don't recall if it handles device-tree child nodes (at all - or properly).
so we need someone to test the result of this series (there was a big Home Hub 5A community - not sure if that's still the case)

neheb replied:
Oh that is very unfortunate.

neheb replied:
ping @alexceltare2 @sch-m @jaimet

xdarklight commented:
> @DragonBluep it's claimed to use AR9287 but the PCI ID claims AR9227. I assume they are equivalent.

to my knowledge AR9227 and AR9287 only differ in the bus they connect to:
- AR9227: PCI (no "e" at the end)
- AR9287: PCIe

neheb commented:
I don’t have any of these devices. Is it possible to verify calibration sizes?

alexceltare2 commented:
I can test on mine. Where do i grab the sysupgrade?

jaimet commented:
I have several spare v5a devices sitting around, so I can also test if there is a binary that I can flash. HTH.

neheb commented:
enjoy: https://mega.nz/file/6JJxmCya#x3W6hRDEuAMzo5eLgBdskt9C8iHjUTK3clNwHTRFZ7s

I added luci for ease.

jaimet commented:
> enjoy: https://mega.nz/file/6JJxmCya#x3W6hRDEuAMzo5eLgBdskt9C8iHjUTK3clNwHTRFZ7s

Thank you for making the binary. I've flashed it, but after firstboot, `/etc/config/wireless` is empty (length 0). I have attached the dmesg output in case this helps. If you specific questions / requests, please let me know. I may be able to provide you with a login to this router, but this will take longer to organise. HTH.

[dmesg.output.txt](https://github.com/user-attachments/files/18230198/dmesg.output.txt)

neheb commented:
@jaimet are the mac addresses on the ethernet interfaces correct?

neheb commented:
try

https://mega.nz/file/iIohhBJS#SVLi_QrMM-VdZMTyWH5GQ3w5AIoy3NqJTGjFfzj0Hec

jaimet commented:
> try
> 
> https://mega.nz/file/iIohhBJS#SVLi_QrMM-VdZMTyWH5GQ3w5AIoy3NqJTGjFfzj0Hec

Interesting.

Regarding mac addresses:
The label mac ends in `25:B0`
The `wan@eth0` device mac matches the label mac except for the last byte (`25:B1` instead of `25:B0`)
All other devices (which are all ethernet) share the same randomly-assigned mac address which changes with every reboot.

Regarding wifi phys, `iw phy` shows `phy1` (2.4GHz) but not `phy0` (5.0GHz). `/etc/config/wireless` is populated accordingly.

I attach the [output from `dmesg`](https://github.com/user-attachments/files/18234154/dmesgOutput.txt) and the [output from `iw phy`](https://github.com/user-attachments/files/18234155/iw.phy-output.txt).

neheb commented:
```
[   25.198595] ath10k_pci 0000:02:00.0: otp calibration failed: 2
```

which means the node is wrong.

Here's a version with lspci:

https://mega.nz/file/bIwgBaaK#YlAJDtSsvMhYK6Y_29ZMpVTymryVyiqqt4OqmIJyZak

I think

lspci
lspci -t

would be useful.

As for the ethernet mac being wrong and only WAN being correct, only thing I can think of is the ethernet driver not supporting EPROBE_DEFER properly.

neheb commented:
Alright this should finally work

https://mega.nz/file/zBhwXZIQ#W1Kl1TRzWQo1sSQ3gE_vY4Wj64Taz-_DEuKw9Wrtw-c

notes:

As @xdarklight said, this platform uses a weird pcie configuration where the pcie device is under a bridge. Should work now.

I put the nvmem stuff under the switch node instead of ethernet. That has been fixed.

jaimet commented:
> I think
> lspci lspci -t
> would be useful.

```
# lspci
00:0e.0 Network controller: Qualcomm Atheros AR9227 Wireless Network Adapter (rev 01)
01:00.0 PCI bridge: Lantiq MIPS SoC PCI Express Port (rev 01)
02:00.0 Network controller: Qualcomm Atheros QCA986x/988x 802.11ac Wireless Network Adapter
```

```
# lspci -t
-+-[0000:00]---0e.0
 \-[0000:01]---00.0-[02]----00.0
```

I haven't included the other outputs because I presume that the only change is the addition of `lspci`. Please let me know if I am wrong / if you want any other output. I'll try the [next one](https://github.com/openwrt/openwrt/pull/17289#issuecomment-2560497224) now...

jaimet commented:
> Alright this should finally work
> 
> https://mega.nz/file/zBhwXZIQ#W1Kl1TRzWQo1sSQ3gE_vY4Wj64Taz-_DEuKw9Wrtw-c
> 
> notes:
> 
> As @xdarklight said, this platform uses a weird pcie configuration where the pcie device is under a bridge. Should work now.
> 
> I put the nvmem stuff under the switch node instead of ethernet. That has been fixed.

All good. Both wireless phys detected and functioning, and all mac addresses present and correct. :partying_face: :1st_place_medal: :clap:

neheb commented:
ping @robimarko

neheb commented:
ping @abajk 

I believe you have this unit.

abajk commented:
> Rebased. @abajk does this work?

I will test this PR on Saturday.

abajk commented:
@neheb Finally, I was able to test your patch. Everything matches except the DSL mac address.

~~~
OpenWRT version:
eth0	some random value? (56:4C:45:08:37:76)
lan	label
wan	same as eth0
dsl	label+1
wifi2	label+2
wifi5	label+3

neheb version:
eth0	label
lan	label
wan	label+1
dsl 	some random value? (00:20:DA:86:23:75)
wifi2	label+2
wifi5	label+3
~~~

abajk commented:
I restored the stock software. Below is the vendor's MAC setup:
~~~
lan	label+0
wan	label+4
dsl	label+5
wifi2	label+2
wifi5	label+3
~~~
I can't find where label+1 address is used.

jaimet commented:
> @neheb Finally, I was able to test your patch. Everything matches except the DSL mac address.
> 
> ```
> OpenWRT version:
> eth0	some random value? (56:4C:45:08:37:76)
> lan	label
> wan	same as eth0
> dsl	label+1
> wifi2	label+2
> wifi5	label+3
> 
> neheb version:
> eth0	label
> lan	label
> wan	label+1
> dsl 	some random value? (00:20:DA:86:23:75)
> wifi2	label+2
> wifi5	label+3
> ```

Interesting. I'm running release 23.05.4 (r24012) and my macs match what you have written for the neheb version, not what you have written for the openwrt version. Just to confirm:

```
OpenWRT release 23.05.4 (r24012):
eth0:      label
lan:       label
wan:       label+1
phy2.4:    label+2
phy5.0:    label+3

```
I don't know whether this makes a difference, but as I have FTTP, I am not using the DSL port at all.

abajk commented:
OpenWRT sets the MAC address on the currently used WAN interface. In your case you see this address on the WAN port, in my case it is the DSL port. The standard software uses two different addresses here.

abajk commented:
After the changes from neheb, the MAC addresses are registered in the dts (lan, wan, wifi2.4, wifi5). The dsl interface is not created based on the DTS file. It is created dynamically after the link is established with the DSLAM.

neheb commented:
In other words the DSL MAC still needs to be set in user space. Unfortunate.

Anyway, should this be redone to match stock firmware behavior?

abajk commented:
My provider migrated to FTTH. I no longer use VDSL.
Maybe @janh  will know if a random MAC address can cause some problems in DSLAM.

neheb commented:
Ideally

```
--- a/target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_bt_homehub-v5a.dts
+++ b/target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_bt_homehub-v5a.dts
@@ -216,11 +216,16 @@
                rx-internal-delay-ps = <1500>;
                phy-handle = <&phy5>;
 
-               nvmem-cells = <&macaddr_caldata_110c 1>;
+               nvmem-cells = <&macaddr_caldata_110c 4>;
                nvmem-cell-names = "mac-address";
        };
 };
 
+&vdsl {
+       nvmem-cells = <&macaddr_caldata_110c 5>;
+       nvmem-cell-names = "mac-address";
+};
+
 &localbus {
        flash@1 {
                compatible = "lantiq,nand-xway";

```

would do it I think.

neheb committed:
ltq-ptm: add NVMEM MAC support

This will be used in the following commit to move MAC assignment of the
DSL interface to NVMEM.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17289
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb commented:
@abajk @jaimet please retest.

I edited the WAN/DSL assignments to 4/5 as in stock firmware. I also added a patch to get the MAC from NVMEM for (I think?) the DSL interface.

janh commented:
> Maybe @janh will know if a random MAC address can cause some problems in DSLAM.

The DSL MAC address is used as source address for the vectoring error reports. These are generated by the modem firmware and sent to the vectoring engine in the DSLAM. (On VR9 modems they are forwarded by the main CPU, so this has the nice side effect that we can inspect them easily. The filter `llc and ether[14:4]=0xaaaa0300 and ether[18:4]=0x19a70003` should work with tcpdump.)

If unconfigured, the modem firmware uses an all-zero MAC address, and at least on my line that didn't seem to cause any issues. But of course that does not mean it is fine.

The vectoring spec ([ITU-T G.993.5](https://www.itu.int/rec/T-REC-G.993.5/en)) says the following:
> – Source MAC address shall be the MAC address of the VTU-R;

I can't really tell how important it is. Identification should not be an issue, as the actual payload contains a line ID. But in any case, it means that not setting the MAC address correctly would be a violation of the spec.

@neheb Configuration of DSL MAC address happens in the [DSL init script](https://github.com/openwrt/openwrt/blob/main/package/network/config/ltq-vdsl-vr9-app/files/dsl_control). Currently, it gets the MAC address from the network config file (if a "device" section for "dsl0" exists).

So with this change, that is no longer going to work. And the actual "dsl0" interface does not exist yet, when the init script runs (because it is only created during the first DSL connection). However, configuration of the MAC address needs to happen already when the modem is initialized, because it doesn't accept later changes.

If it turns out to be infeasible to set the correct MAC address in the end, I think we should at least generate a proper random one, to make sure the modem does not use an all-zero address.

Unfortunately, I cannot test this particular change myself, as I don't have a HH5A. My only xrx200/vr9 devices are a Fritzbox 7362 SL and 7412.

neheb commented:
@janh notice @jaimet 's MAC address for the DSL interface:

```
dsl 	some random value? (00:20:DA:86:23:75)
```

That is exactly the value set in the driver. I would think the code now takes the MAC from NVMEM and sets that.

janh commented:
@neheb The MAC address of the DSL network interface is not the same as the that is used by the firmware for vectoring error reports. The latter one always needs to be explicitly configured (as done by the init script now).

jaimet commented:
> @janh notice @jaimet 's MAC address for the DSL interface:
> ```
> dsl 	some random value? (00:20:DA:86:23:75)
> ```

Minor correction - this was @abajk's data - I was only quoting it.

I am happy to test anything but could you please upload the binaries to mega. Thank you.

abajk commented on target/linux/lantiq/files/arch/mips/boot/dts/lantiq/vr9_bt_homehub-v5a.dts:
```diff
@@ -210,9 +215,17 @@
 	};
 };
 
+&mei {
```
Build fails on:
~~~
Error: /home/aleksander/workspace/openwrt/build_dir/target-mips_24kc_musl/linux-lantiq_xrx200/linux-6.6.74/arch/mips/boot/dts/lantiq/vr9_bt_homehub-v5a.dts:224.1-5 Label or path mei not found
FATAL ERROR: Syntax error parsing input tree
~~~

abajk commented:
I have rebuilt the image and I still see this MAC address on the DSL interface:
~~~
dsl0      Link encap:Ethernet  HWaddr 00:20:DA:86:23:75  
          inet6 addr: fe80::220:daff:fe86:2375/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:91 errors:0 dropped:0 overruns:0 frame:0
          TX packets:134 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:13836 (13.5 KiB)  TX bytes:14758 (14.4 KiB)
~~~

neheb commented:
hrm unfortunate. I'm guessing -EPROBE_DEFER is being returned and that stock address is being set. Added some code handling that.

abajk commented:
Unfortunately, there is still no success. Still have the address ‘00:20:DA:86:23:75’ on the dsl0 interface.

neheb commented:
@abajk unbelievable. I had the wrong node set. If it doesn't work now I'm out of ideas.

Meaning the only thing left is probably to fix the DSL script if applicable.

janh commented:
Lines using ATM instead of PTM might be another issue.

neheb commented:
The simplest solution honestly is to revert to userspace assignment (while keeping NVMEM) for compatibility with the init script.

neheb commented:
The current version doesn’t have the right MAC? I set the right node this time

neheb commented:
@abajk rebased. I'm not sure this will work correctly.

abajk commented:
> @abajk rebased. I'm not sure this will work correctly.

I will test the current version tomorrow. At first look it seems ok.

abajk commented:
It works fine.
~~~
Tested-by: Aleksander Jan Bajkowski <olek2@wp.pl>
~~~

neheb commented:
ping @hauke

neheb commented:
this was successfully tested. I assume this can go in.

neheb commented:
ping

neheb commented:
Rebased just in case.

ping @robimarko

neheb commented:
ping

neheb commented:
I guess this will never get merged.

robimarko commented:
Can you rebase this for 6.12?

I will merge it then since everybody has had a chance to test

neheb commented:
Rebased.

robimarko commented:
Thanks! Rebased on top of main and merged!

neheb commented:
I don’t see why not
